### PR TITLE
Stop `get_joined_users` corruption from custom statuses

### LINKED ISSUE #7099: Messages not being sent over federation
### Description

I'll try to do this as exact as I'm able to in reverse. 

After upgrading our matrix server running on ubuntu 18.04 to 1.11.1+bionic1 we did not receive any messages from servers running on older instances (only in 1:1 chats, groups have been working fine).

We are explicitly talking with a another company running a matrix instance and a freelancer and his own setup.

Both problems were solved after upgrading their versions to the matrix synapse 1.11.1 version. But since at least one party was running 1.11.0+bionic1 before, it looks like the bug has been introduced in the patch release.

All messages and presence updates going from 1.11.1+bionic1 to the other sides went through without problems. But nothing the other way around.

One major pain point was that we discovered this by getting suspicious that some people did not respond in time and thus tried other communication channels. So the users where totally unaware of this.

A last note: As mentioned bringing all instances to 1.11.1 (or the latest docker image) solved it, no message went missing! As soon as the other parties wrote a new message all the missing ones got redelivered.

**Edit**: As far as I can tell it did not matter on which side the 1:1 chatroom has been initially created.

### Steps to reproduce

(I did not re-test this, but this should result in the problematic setup)
- Connect two matrix servers running 1.11.0+bionic1
- Create two 1:1 chats, one initialized by each side
- Upgrade on instance to 1.11.1+bionic1

You should now only be able to send 1:1 messages from 1.11.1+bionic1 to 1.11.0+bionic1 but not the other way around (At least using any riot the user won't get any feedback about this. I'll write an issue for that project later).

I could not find any suspicious log entries on our 1.11.1 server. But I've requested logs from the other side, maybe we can find something there.

### Version information

- apt package installed: matrix-synapse-py3                    1.11.1+bionic1                                  amd64
- Not exactly sure about the other sides. One had matrix-synapse-py3 1.11.0+bionic1 apt
- The other party had the docker container setup using matrixdotorg/synapse (version before upgrade not known)

### PR DESCRIPTION
**Pull request by richvdh**
Fix a bug where the `get_joined_users` cache could be corrupted by custom
status events (or other state events with a state_key matching the user ID).

The bug was introduced by #2229, but has largely gone unnoticed since then.

Fixes #7099, #7373.

### CODE REVIEW

**Comment by richvdh** [comment]
TODO: write a test for this.

**Comment by richvdh** [comment]
This code is trying to figure out who is in the room at point X, which it does by starting with the membership list at point Y, clearing the membership state for any users who changed membership between Y and X, and then adding back in any users whose membership changes were join events.

The bug is that the "clearing the membership state" bit is not correctly limited to membership events.

**Comment by Sorunome** [comment]
This should also fix #7373

**Comment by ErrorProne** [comment]
Do you guys know if this will make it into the 1.13.0 version?

**Comment by richvdh** [comment]
oh balls. No, because it hasn't been merged yet due to the absence of regression tests, and the release branch has been cut :/

**Comment by richvdh** [comment]
It looks like we're going to have to do another RC anyway. Let's try to get this one in.

**Comment by clokep** [comment]
@richvdh I merged #7489 so might want to merge `release-v1.13.0` in to fix the build.

File: changelog.d/7376.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix a bug which could cause messages not to be sent over federation, when state events with state keys matching user IDs (such as custom user statuses) were received statuses.
```
Comment (anoadragon453) [inline changelog.d/7376.bugfix:1]:
```suggestion
Fix a bug which could cause messages not to be sent over federation, when state events with state keys matching user IDs (such as custom user statuses) were received.
```

File: tests/storage/test_roommember.py [hunk 2]
```diff
@@ -114,6 +116,52 @@ def test_count_known_servers_stat_counter_enabled(self):
...
+            type="m.test.1",
+            content={},
+        )
+
+        users = self.get_success(
+            self.store.get_joined_users_from_context(event, context)
+        )
+        self.assertEqual(users.keys(), {self.u_alice, self.u_bob})
+
+        # Regression test for #7376: create a state events whose key matches bob's
```
Comment (anoadragon453) [inline tests/storage/test_roommember.py:140]:
```suggestion
        # Regression test for #7376: create a state event whose key matches bob's
```

File: tests/storage/test_roommember.py [hunk 3]
```diff
@@ -114,6 +116,52 @@ def test_count_known_servers_stat_counter_enabled(self):
...
+        # user_id, but which is *not* a membership event, and persist that; then check
+        # what the state is for the next event.
```
Comment (anoadragon453) [inline tests/storage/test_roommember.py:142]:
```suggestion
        # whether we can still successfully query joined room members.
```

May be more clear?

Comment (richvdh) [inline tests/storage/test_roommember.py:142]:
```suggestion
        # that `get_joined_users_from_context` returns the correct users for the next event.
```

**Comment by richvdh** [comment]
thanks for the suggestions @anoadragon453, all good spots!

**Pull request merged**